trip_app/trips/views.py

```diff
@@ -69,35 +80,79 @@ def get_queryset(self):
 
     @action(detail=True, methods=["post"])
     def reserve(self, request, pk=None):
-        """ Reserve a trip by sending a POST to api/trip/<trip_id>/reserve/."""
+        """ Reserve a trip via POST api/trips/<trip_id>/reserve/. """
         trip = self.get_object()
         user = request.user
         if user == trip.driver:
             return Response(
                 data="Driver can't be a passenger at the same time",
                 status=status.HTTP_400_BAD_REQUEST,
             )
-        if trip.free_seats:
-            if trip.man_approve:
-                trip_request = TripRequest.objects.create(trip=trip, user=user)
-                serializer = TripRequestSerializer(trip_request)
-                return Response(
-                    serializer.data, status=status.HTTP_201_CREATED
-                )
-            trip.passengers.add(user)
-            serializer = self.get_serializer(trip)
-            return Response(serializer.data)
-        return Response(
-            data="There are no empty seats in this trip",
-            status=status.HTTP_400_BAD_REQUEST,
-        )
-
-    @action(
-        detail=True, methods=["get"], permission_classes=[IsTripDriverOrAdmin]
-    )
+        if user.requests.filter(trip_id=pk).exists():
+            return Response(
+                data="You have already requested this trip",
+                status=status.HTTP_400_BAD_REQUEST,
+            )
```

Comment:
These don't feel to belong to the ViewSet level as well. What do you think?

Reply:
Yeah... I have a bad feeling that it is not correct place for them as well.. just don't know where to replace them. One option that I have in mind is to create custom permission for this view, but how do you think is this correct to return 403 Forbidden or so for such cases ?

Reply:
Added new permission class for this view to embed this logic.